File: Cloud Functions/functions/src/RecommendItems.js
const MIN_SUPPORT_THRESH = 0;

class Rule {
    constructor(prior, antecedent) {
        this.prior = prior;
        this.antecedent = antecedent;
    }

    getPrior() {
        return(this.prior);
    }

    getAntecedent() {
        return(this.antecedent);
    }
}

class Node {
    constructor(item) {
        this.item = item;
        this.count = 1;
        this.children = [];
    }

    addChild(childItem) {
        var newChild = new Node(childItem);
        this.children.push(newChild);
        return(newChild);
    }
}

class FPTree {
    constructor(){
        this.root = null;
    }

    addPath(listOfItems){
        var currNode = this.root;
        while (listOfItems.length > 0) {
            var currVal = listOfItems[0];

            var ind = currNode.children.indexOf(currVal)
            if (ind === -1) {
                currNode = currNode.addChild(currVal);
            } else {
                currNode.incCount();
                currNode = currNode.children[ind];
            }
            
            listOfItems.splice(0, 1);
        }
    }
}

function calcSupport(itemSubset, allTransactions) {
    var numIncludes = 0;

    for (var i = 0; i < allTransactions.length; i++) {
        var currTransaction = allTransactions[i];
        var success = itemSubset.every(val => currTransaction.includes(val));
        if (success) {
            numIncludes += 1;
        }
    }

    var support = numIncludes / allTransactions.length;

    return(support);
}

function calcConfidence(ruleToCheck, allTransactions) {
    var prior = ruleToCheck.getPrior();
    var antecedent = ruleToCheck.getAntecedent();

    var union = antecedent.concat(prior);

    var numerator = calcSupport(union, allTransactions);
    var denominator = calcSupport(prior, allTransactions);

    var confidence = numerator / denominator;

    return(confidence);
}

function constructFPTree(allTransactions) {
    var supportMap = {};

    for (var i = 0; i < allTransactions.length; i++) {
        var currTransaction = allTransactions[i];
        for (var j = 0; j < currTransaction.length; j++) {
            var item = currTransaction[j];

            if (!(item in supportMap)) {
                var supp = calcSupport(item, allTransactions);

                supportMap[item] = supp;
            }
        }
    }

    var descOrder = [];
    for (var item in supportMap) {
        descOrder.push([item, supportMap[item]]);
    }

    descOrder.sort(function(a, b) {
        return (a[1] < b[1]) ? -1 : (a[1] > b[1]) ? 1 : 0;
    });

    var finalOrder = [];
    for (var i = 0; i < descOrder.length; i++) {
        finalOrder.push(descOrder[i]);
    }

    for (var i = 0; i < allTransactions.length; i++) {
        var currTransaction = allTransactions[i];
        
        currTransaction.sort(function(a, b) {
            var indA = finalOrder.indexOf(a);
            var indB = finalOrder.indexOf(b);

            return (indA < indB) ? -1 : (indA > indB) ? 1 : 0;
        });
    }
}

File: gcf/RecommendItemFiles/RecommendItems.py
from FPTreeObj import FPTree
from copy import copy

from itertools import chain, combinations

def calcSupport(itemSubset, allTransactions):
    numIncludes = 0

    for currTransaction in allTransactions:
        if (itemSubset.issubset(currTransaction)):
            numIncludes += 1

    support = float(numIncludes) / len(allTransactions)

    return support

def calcConfidence(ruleToCheck, allTransactions):
    prior = ruleToCheck.getPrior()
    antecedent = ruleToCheck.getAntecedent()

    union = antecedent.union(prior)

    numerator = calcSupport(union, allTransactions)
    denominator = calcSupport(prior, allTransactions)

    confidence = numerator / denominator

    return confidence

def constructFPtree(allTransactions):
    supportMap = {}

    for currTransaction in allTransactions:
        for item in currTransaction:
            if item not in supportMap:
                supp = calcSupport(set(item), allTransactions)
                supportMap[item] = supp

    descOrder = list(reversed(sorted(supportMap, key=lambda item: supportMap[item])))

    descOrder = ["B", "E", "A", "C", "D"]

    tree = FPTree()

    for currTransaction in allTransactions:
        sortedItem = sorted(currTransaction, key=lambda item: descOrder.index(item))

        tree.addPath(sortedItem)

    return(tree)

def powerset(iterable):
    s = list(iterable)
    return chain.from_iterable(combinations(s, r) for r in range(1, len(s)+1))

def mineTree(tree, root = True, currPath = None):
    paths = {}

    item = tree.getItem()

    if (currPath == None):
        currPath = []
        temp = currPath
    else:
        temp = copy(currPath)
        temp.append(item)

    if (root == False):
        if (item not in paths):
            paths[item] = []

        paths[item].append((tree.getCount(), currPath))

    children = tree.getChildren()
    for child in children:
        tempPaths = mineTree(child, root=False, currPath=temp)
        for item in tempPaths:
            if item in paths:
                paths[item] += tempPaths[item]
            else:
                paths[item] = tempPaths[item]

    if (root == False):
        return(paths)
    else:
        rules = set()
        for item in paths:
            for currList in paths[item]:
                currPowerset = list(powerset(currList[1]))
                currPowerset = [list(x) for x in currPowerset]
                currRuleSet = [[item] + x for x in currPowerset]
                rules.update(set(tuple(i) for i in currRuleSet))
        for rule in rules:
            print(rule)


def formatTransactions(allTransactions):
    newTransactions = []
    for transaction in allTransactions:
        newTransactions.append(set(transaction))
    return(newTransactions)

if __name__ == "__main__":
    transactions = [["B", "E", "A", "D"],
                    ["B", "C", "E"],
                    ["A", "B", "D", "E"],
                    ["A", "B", "C", "E"],
                    ["A", "B", "C", "D", "E"],
                    ["B", "C", "D"]]

    transactions = formatTransactions(transactions)

    tree = constructFPtree(transactions)

    mineTree(tree)